File: ixian-mission-critical/src/lib/schemas/task.ts
/**
 * Task validation schemas
 */

import { z } from 'zod';
import { RECURRENCE_OPTIONS } from '$lib/db';

/**
 * Task tag schema (now accepts any string for custom tags)
 */
export const taskTagSchema = z.string().min(1, 'Tag is required').max(50, 'Tag name is too long');

/**
 * Recurrence type enum schema
 */
export const recurrenceSchema = z.enum(RECURRENCE_OPTIONS);

/**
 * Schema for creating a new task
 */
export const createTaskSchema = z.object({
	text: z.string().min(1, 'Task title is required').max(500, 'Task title is too long'),
	description: z.string().max(2000, 'Description is too long').default(''),
	important: z.boolean().default(false),
	tag: taskTagSchema.default('General'),
	dueAt: z.number().nullable().default(null),
	recurrence: recurrenceSchema.default('none'),
	recurrenceAlt: z.boolean().default(false)
});

export type CreateTaskInput = z.infer<typeof createTaskSchema>;

/**
 * Schema for updating a task
 */
export const updateTaskSchema = z.object({
	text: z.string().min(1, 'Task title is required').max(500, 'Task title is too long').optional(),
	description: z.string().max(2000, 'Description is too long').optional(),
	completed: z.boolean().optional(),
	important: z.boolean().optional(),
	tag: taskTagSchema.optional(),
	dueAt: z.number().nullable().optional(),
	recurrence: recurrenceSchema.optional(),
	recurrenceAlt: z.boolean().optional()
});

export type UpdateTaskInput = z.infer<typeof updateTaskSchema>;

/**
 * Full task schema (for validation from server)
 */
export const taskSchema = z.object({
	id: z.string().uuid(),
	text: z.string(),
	description: z.string(),
	completed: z.boolean(),
	important: z.boolean(),
	tag: taskTagSchema,
	dueAt: z.number().nullable(),
	recurrence: recurrenceSchema,
	recurrenceAlt: z.boolean(),
	createdAt: z.number(),
	updatedAt: z.number(),
	deletedAt: z.number().nullable()
});

export type TaskData = z.infer<typeof taskSchema>;

File: ixian-mission-critical-backend/app/schemas/task.py
"""
Pydantic schemas for Task API.
"""
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class RecurrenceType(str, Enum):
    """Valid recurrence types."""

    NONE = "none"
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"


class TaskBase(BaseModel):
    """Base schema with common task fields."""

    text: str = Field(..., min_length=1, max_length=500)
    description: str = Field(default="", max_length=2000)
    completed: bool = False
    important: bool = False
    tag: str = Field(default="General", max_length=50)  # Now dynamic - references tag name
    due_at: int | None = None  # Unix timestamp (ms)
    recurrence: RecurrenceType = RecurrenceType.NONE
    recurrence_alt: bool = False


class TaskCreate(TaskBase):
    """Schema for creating a new task. Client may supply id and timestamps for sync."""

    id: UUID | None = None
    created_at: int | None = Field(default=None, ge=0)
    updated_at: int | None = Field(default=None, ge=0)
    deleted_at: int | None = Field(default=None, ge=0)


class TaskUpdate(BaseModel):
    """Schema for updating an existing task. All fields optional."""

    text: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=2000)
    completed: bool | None = None
    important: bool | None = None
    tag: str | None = Field(default=None, max_length=50)
    due_at: int | None = None
    recurrence: RecurrenceType | None = None
    recurrence_alt: bool | None = None
    created_at: int | None = Field(default=None, ge=0)
    updated_at: int | None = Field(default=None, ge=0)
    deleted_at: int | None = Field(default=None, ge=0)


class TaskResponse(TaskBase):
    """Schema for task responses including server-generated fields."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: int  # Unix timestamp (ms)
    updated_at: int  # Unix timestamp (ms)
    deleted_at: int | None = None
